pkg/cloud/aliyunprovider.go

```diff
@@ -0,0 +1,828 @@
+func (aliyun *Aliyun) AllNodePricing() (interface{}, error) {
+	aliyun.DownloadPricingDataLock.RLock()
+	defer aliyun.DownloadPricingDataLock.RUnlock()
+	return aliyun.Pricing, nil
+}
+
+// NodePricing gives a specific node for the key
+func (aliyun *Aliyun) NodePricing(key Key) (*Node, error) {
+	aliyun.DownloadPricingDataLock.RLock()
+	defer aliyun.DownloadPricingDataLock.RUnlock()
+
+	// Get node features for the key
+	keyFeature := key.Features()
+
+	pricing, ok := aliyun.Pricing[keyFeature]
+	if !ok {
+		log.Infof("Node pricing information not found for node with feature: %s", keyFeature)
+		return &Node{}, nil
+	}
+
+	log.Infof("returing the node price for the node with feature: %s", keyFeature)
+	return pricing.Node, nil
+}
+
+// PVPricing gives a specific PV price for the PVkey
+func (aliyun *Aliyun) PVPricing(pvk PVKey) (*PV, error) {
+	aliyun.DownloadPricingDataLock.RLock()
+	defer aliyun.DownloadPricingDataLock.RUnlock()
+
+	keyFeature := pvk.Features()
+
+	pricing, ok := aliyun.Pricing[keyFeature]
+
+	if !ok {
+		log.Infof("Persistent Volume pricing not found for PV with feature: %s", keyFeature)
+		return &PV{}, nil
+	}
+
+	log.Infof("returing the PV price for the node with feature: %s", keyFeature)
+	return pricing.PV, nil
+}
+
+// Stubbed NetworkPricing for Aliyun. Will look at this in Next PR
+func (aliyun *Aliyun) NetworkPricing() (*Network, error) {
+	return &Network{
+		ZoneNetworkEgressCost:     0.0,
+		RegionNetworkEgressCost:   0.0,
+		InternetNetworkEgressCost: 0.0,
+	}, nil
+}
+
+// Stubbed LoadBalancerPricing for Aliyun. Will look at this in Next PR
+func (aliyun *Aliyun) LoadBalancerPricing() (*LoadBalancer, error) {
+	return &LoadBalancer{
+		Cost: 0.0,
+	}, nil
+}
+
+func (aliyun *Aliyun) GetConfig() (*CustomPricing, error) {
+	c, err := aliyun.Config.GetCustomPricingData()
+	if err != nil {
+		return nil, err
+	}
+	if c.Discount == "" {
+		c.Discount = "0%"
+	}
+	if c.NegotiatedDiscount == "" {
+		c.NegotiatedDiscount = "0%"
+	}
+	if c.ShareTenancyCosts == "" {
+		c.ShareTenancyCosts = defaultShareTenancyCost
+	}
+
+	return c, nil
+}
+
+// Load once and cache the result (even on failure). This is an install time secret, so
+// we don't expect the secret to change. If it does, however, we can force reload using
+// the input parameter.
+func (aliyun *Aliyun) loadAliyunAuthSecretAndSetEnv(force bool) error {
+	if !force && aliyun.loadedAccessKey {
+		return nil
+	}
+
+	exists, err := fileutil.FileExists(authSecretPath)
+	if !exists || err != nil {
+		return fmt.Errorf("Failed to locate service account file: %s", authSecretPath)
+	}
+
+	result, err := ioutil.ReadFile(authSecretPath)
+	if err != nil {
+		return err
+	}
+
+	var ak *AliyunAccessKey
+	err = json.Unmarshal(result, &ak)
+	if err != nil {
+		return err
+	}
+
+	err = env.Set(env.AliyunAccessKeyIDEnvVar, ak.AccessKeyID)
+	if err != nil {
+		return err
+	}
+	err = env.Set(env.AliyunAccessKeySecretEnvVar, ak.SecretAccessKey)
+	if err != nil {
+		return err
+	}
```

Comment:
Setting environment variables mid-run of a program and planning to read them back later is a really bizarre pattern. I'm not sure if I've ever seen it before -- it bothers me because the environment becomes a mutable thing and thus very difficult to reason about.

This may be mimicking other provider code, though...